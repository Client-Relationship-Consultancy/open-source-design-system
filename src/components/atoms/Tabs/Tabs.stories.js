import React from "react"
import { storiesOf } from "@storybook/react"
import { text, boolean } from "@storybook/addon-knobs"

import Tabs from "./Tabs"

const stories = storiesOf("Tabs", module)

stories.add("Tabs", () => {
  return (
    <div style={{ paddingBottom: "5rem", width: "400px" }}>
      <Tabs
        tabs={[
          {
            id: "tab1",
            header: "activity log",
            content: <div>sure!</div>,
            default: true,
          },
          {
            id: "tab2",
            header: "activity log 1",
            content: <div>wooow</div>,
          },
        ]}
      />
    </div>
  )
})
