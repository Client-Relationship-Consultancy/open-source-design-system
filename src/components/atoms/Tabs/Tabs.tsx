import React from "react";
import styled from "styled-components";
import { colourPalette } from "../../../brandColours";

const StyledTabs = styled.div`
  display: flex;
  position:sticky;
  top:0;
  background:${(props) => props.theme.surface.hex};
  flex-direction: row;
  color: ${(props) => props.theme.black.main.hex};
  font-size: 1rem;
  font-family: "Gentona", "Montserrat";
  width: 100%;
  align-items: stretch;
  button {
    line-height: 22px;
    font-size: 14px;
    text-align: center;
    flex-grow: 1;
    background: transparent;
    border: none;
    outline-width: 0;
    padding: 7.5px 0;
  }
  button::hover {
    outline-width: 0;
  }
  button.selected {
    border-bottom: 2px solid ${(props) => props.theme.action.main.hex};
    font-weight: 600;
    color: ${(props) => props.theme.action.main.hex};
  }
`;

StyledTabs.defaultProps = {
  theme: colourPalette.examplePalette,
};

StyledTabs.displayName = "StyledTabs";

// type TabsProps = React.HTMLProps<HTMLTabsElement>;
interface IState {
  selected: string;
}

interface ITab {
  id: string;
  header: string;
  content: React.ReactNode | React.ReactElement;
  default?: boolean;
}

interface IProps {
  tabs: ITab[];
}

function getDefaultTabId(tabs: ITab[]): string {
  const defaultTab = tabs.filter((tab) => tab.default);
  if (defaultTab.length > 0) {
    return defaultTab[0].id;
  }
  return tabs[0].id;
}

class Tabs extends React.Component<IProps, IState> {
  state = {
    selected: getDefaultTabId(this.props.tabs),
  };

  getSelectedTabContent = (): React.ReactNode | undefined => {
    const selectedTab = this.props.tabs.filter((tab) => tab.id === this.state.selected);
    if (selectedTab.length > 0) {
      return selectedTab[0].content;
    }
    return undefined;
  };

  selectTab = (id: string): void => {
    this.setState({ selected: id });
  };

  render() {
    return (
      <>
        <StyledTabs>
          {this.props.tabs.map((tab) => (
            <button
              className={this.state.selected === tab.id ? "selected" : undefined}
              onClick={this.selectTab.bind(this, tab.id)}
              type="button"
              key={tab.id}
            >
              {tab.header}
            </button>
          ))}
        </StyledTabs>
        {this.state.selected && this.getSelectedTabContent()}
      </>
    );
  }
}

export default Tabs;
