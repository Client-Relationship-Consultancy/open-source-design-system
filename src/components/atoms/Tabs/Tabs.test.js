import React from "react"
import { mount, shallow } from "enzyme"
import toJson from "enzyme-to-json"
import "jest-styled-components"
import Tabs, { StyledTabs } from "./Tabs"

describe("Tabs", () => {
  const props = {
    tabs: [
      {
        id: "tab1",
        header: "activity log",
        content: <span>sure!</span>,
        default: true,
      },
      {
        id: "tab2",
        header: "activity log 1",
        content: <span>wooow</span>,
      },
    ],
  }
  const component = shallow(<Tabs {...props} />)

  it("should match the snapshot", () => {
    const tree = toJson(component)
    expect(tree).toMatchSnapshot()
  })

  it("StyledTabs should match the snapshot", () => {
    const StyledTabsShallow = mount(
      <StyledTabs
        theme={{
          surface: { hex: "mockColorsurface" },
          black: { main: { hex: "mockColorblack" } },
          action: { main: { hex: "mockColoraction" } },
        }}
      >
        <button className="selected" type="button">
          tab
        </button>
      </StyledTabs>,
    )
    const tree = toJson(StyledTabsShallow)
    expect(tree).toMatchSnapshot()
  })

  it("render the default tab", () => {
    expect(component.find("button.selected").text()).toEqual("activity log")
    expect(component.find("span").text()).toEqual("sure!")
  })
  it("render the selected tab", () => {
    component.instance().selectTab("tab2")
    expect(component.find("button.selected").text()).toEqual("activity log 1")
    expect(component.find("span").text()).toEqual("wooow")
  })

  it("getSelectedTabContent return undefined if no tab is selected", () => {
    component.setState({ selected: "not available" })
    expect(component.instance().getSelectedTabContent()).toEqual(undefined)
  })

  it("should return the first tab selected if no tab as default set to true", () => {
    const propsMock = {
      tabs: [
        {
          id: "tab1",
          header: "activity log",
          content: <span>sure!</span>,
        },
        {
          id: "tab2",
          header: "activity log 1",
          content: <span>wooow</span>,
        },
      ],
    }
    const componentMock = shallow(<Tabs {...propsMock} />)
    expect(componentMock.find("button.selected").text()).toEqual("activity log")
    expect(componentMock.find("span").text()).toEqual("sure!")
  })
})
