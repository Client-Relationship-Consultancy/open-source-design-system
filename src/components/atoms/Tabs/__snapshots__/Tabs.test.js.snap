// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs StyledTabs should match the snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: -webkit-sticky;
  position: sticky;
  top: 0;
  background: mockColorsurface;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  color: mockColorblack;
  font-size: 1rem;
  font-family: "Gentona","Montserrat";
  width: 100%;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  z-index: 9999;
}

.c0 button {
  line-height: 22px;
  font-size: 14px;
  text-align: center;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  background: transparent;
  border: none;
  outline-width: 0;
  padding: 7.5px 0;
}

.c0 button::hover {
  outline-width: 0;
}

.c0 button.selected {
  border-bottom: 2px solid mockColoraction;
  font-weight: 600;
  color: mockColoraction;
}

<StyledTabs
  theme={
    Object {
      "action": Object {
        "main": Object {
          "hex": "mockColoraction",
        },
      },
      "black": Object {
        "main": Object {
          "hex": "mockColorblack",
        },
      },
      "surface": Object {
        "hex": "mockColorsurface",
      },
    }
  }
>
  <StyledComponent
    forwardedComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "_foldedDefaultProps": Object {
          "theme": Object {
            "action": Object {
              "dark": Object {
                "RGB": "79,38,98",
                "hex": "#4E2662",
                "on": "#FFFFFF",
              },
              "light": Object {
                "RGB": "173,101,207",
                "hex": "#AD65CF",
                "on": "#FFFFFF",
              },
              "main": Object {
                "RGB": "116,60,143",
                "hex": "#743C8F",
                "on": "#FFFFFF",
              },
            },
            "background": Object {
              "RGB": "246,246,246",
              "hex": "#F6F6F6",
              "on": "#272727",
            },
            "black": Object {
              "main": Object {
                "RGB": "39,39,39",
                "hex": "#272727",
                "on": "#FFFFFF",
              },
              "tint20": Object {
                "RGB": "236,236,236",
                "hex": "#ECECEC",
                "on": "#272727",
              },
              "tint40": Object {
                "RGB": "199,199,199",
                "hex": "#C7C7C7",
                "on": "#272727",
              },
              "tint60": Object {
                "RGB": "160,160,160",
                "hex": "#A0A0A0",
                "on": "#FFFFFF",
              },
              "tint80": Object {
                "RGB": "105,105,105",
                "hex": "#696969",
                "on": "#FFFFFF",
              },
            },
            "complimentary": Object {
              "dark": Object {
                "RGB": "206,185,0",
                "hex": "#CEB900",
                "on": "#272727",
              },
              "light": Object {
                "RGB": "255,245,157",
                "hex": "#FFF59D",
                "on": "#272727",
              },
              "main": Object {
                "RGB": "255,238,82",
                "hex": "#FFEE52",
                "on": "#272727",
              },
            },
            "error": Object {
              "dark": Object {
                "RGB": "197,73,1",
                "hex": "#C54901",
                "on": "#FFFFFF",
              },
              "main": Object {
                "RGB": "255,115,32",
                "hex": "#FF7320",
                "on": "#FFFFFF",
              },
            },
            "primary": Object {
              "dark": Object {
                "RGB": "102,22,20",
                "hex": "#661614",
                "on": "#FFFFFF",
              },
              "light": Object {
                "RGB": "218,52,47",
                "hex": "#DA342F",
                "on": "#FFFFFF",
              },
              "main": Object {
                "RGB": "164,40,36",
                "hex": "#A42824",
                "on": "#FFFFFF",
              },
            },
            "secondary": Object {
              "dark": Object {
                "RGB": "5,38,59",
                "hex": "#05263B",
                "on": "#FFFFFF",
              },
              "light": Object {
                "RGB": "89,156,175",
                "hex": "#599CAF",
                "on": "#FFFFFF",
              },
              "main": Object {
                "RGB": "11,70,107",
                "hex": "#0B466B",
                "on": "#FFFFFF",
              },
            },
            "surface": Object {
              "RGB": "255,255,255",
              "hex": "#FFFFFF",
              "on": "#272727",
            },
            "white": Object {
              "RGB": "255,255,255",
              "hex": "#FFFFFF",
              "on": "#272727",
            },
          },
        },
        "attrs": Array [],
        "componentStyle": ComponentStyle {
          "componentId": "Tabs__StyledTabs-sc-1et8nn2-0",
          "isStatic": false,
          "lastClassName": "c0",
          "rules": Array [
            "display:flex;position:sticky;top:0;background:",
            [Function],
            ";flex-direction:row;color:",
            [Function],
            ";font-size:1rem;font-family:\\"Gentona\\",\\"Montserrat\\";width:100%;align-items:stretch;z-index:9999;button{line-height:22px;font-size:14px;text-align:center;flex-grow:1;background:transparent;border:none;outline-width:0;padding:7.5px 0;}button::hover{outline-width:0;}button.selected{border-bottom:2px solid ",
            [Function],
            ";font-weight:600;color:",
            [Function],
            ";}",
          ],
        },
        "displayName": "StyledTabs",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "Tabs__StyledTabs-sc-1et8nn2-0",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
    theme={
      Object {
        "action": Object {
          "main": Object {
            "hex": "mockColoraction",
          },
        },
        "black": Object {
          "main": Object {
            "hex": "mockColorblack",
          },
        },
        "surface": Object {
          "hex": "mockColorsurface",
        },
      }
    }
  >
    <div
      className="c0"
    >
      <button
        className="selected"
        type="button"
      >
        tab
      </button>
    </div>
  </StyledComponent>
</StyledTabs>
`;

exports[`Tabs should match the snapshot 1`] = `
<Fragment>
  <StyledTabs
    theme={
      Object {
        "action": Object {
          "dark": Object {
            "RGB": "79,38,98",
            "hex": "#4E2662",
            "on": "#FFFFFF",
          },
          "light": Object {
            "RGB": "173,101,207",
            "hex": "#AD65CF",
            "on": "#FFFFFF",
          },
          "main": Object {
            "RGB": "116,60,143",
            "hex": "#743C8F",
            "on": "#FFFFFF",
          },
        },
        "background": Object {
          "RGB": "246,246,246",
          "hex": "#F6F6F6",
          "on": "#272727",
        },
        "black": Object {
          "main": Object {
            "RGB": "39,39,39",
            "hex": "#272727",
            "on": "#FFFFFF",
          },
          "tint20": Object {
            "RGB": "236,236,236",
            "hex": "#ECECEC",
            "on": "#272727",
          },
          "tint40": Object {
            "RGB": "199,199,199",
            "hex": "#C7C7C7",
            "on": "#272727",
          },
          "tint60": Object {
            "RGB": "160,160,160",
            "hex": "#A0A0A0",
            "on": "#FFFFFF",
          },
          "tint80": Object {
            "RGB": "105,105,105",
            "hex": "#696969",
            "on": "#FFFFFF",
          },
        },
        "complimentary": Object {
          "dark": Object {
            "RGB": "206,185,0",
            "hex": "#CEB900",
            "on": "#272727",
          },
          "light": Object {
            "RGB": "255,245,157",
            "hex": "#FFF59D",
            "on": "#272727",
          },
          "main": Object {
            "RGB": "255,238,82",
            "hex": "#FFEE52",
            "on": "#272727",
          },
        },
        "error": Object {
          "dark": Object {
            "RGB": "197,73,1",
            "hex": "#C54901",
            "on": "#FFFFFF",
          },
          "main": Object {
            "RGB": "255,115,32",
            "hex": "#FF7320",
            "on": "#FFFFFF",
          },
        },
        "primary": Object {
          "dark": Object {
            "RGB": "102,22,20",
            "hex": "#661614",
            "on": "#FFFFFF",
          },
          "light": Object {
            "RGB": "218,52,47",
            "hex": "#DA342F",
            "on": "#FFFFFF",
          },
          "main": Object {
            "RGB": "164,40,36",
            "hex": "#A42824",
            "on": "#FFFFFF",
          },
        },
        "secondary": Object {
          "dark": Object {
            "RGB": "5,38,59",
            "hex": "#05263B",
            "on": "#FFFFFF",
          },
          "light": Object {
            "RGB": "89,156,175",
            "hex": "#599CAF",
            "on": "#FFFFFF",
          },
          "main": Object {
            "RGB": "11,70,107",
            "hex": "#0B466B",
            "on": "#FFFFFF",
          },
        },
        "surface": Object {
          "RGB": "255,255,255",
          "hex": "#FFFFFF",
          "on": "#272727",
        },
        "white": Object {
          "RGB": "255,255,255",
          "hex": "#FFFFFF",
          "on": "#272727",
        },
      }
    }
  >
    <button
      className="selected"
      key="tab1"
      onClick={[Function]}
      type="button"
    >
      activity log
    </button>
    <button
      key="tab2"
      onClick={[Function]}
      type="button"
    >
      activity log 1
    </button>
  </StyledTabs>
  <span>
    sure!
  </span>
</Fragment>
`;
